src/server/accept.rs

```diff
@@ -89,83 +74,111 @@ where
-                        let connection = conn::http2::Builder::new(TokioExecutor::new())
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service);
-
-                        // Create a new HTTP/1.1 connection.
-                        #[cfg(all(feature = "http1", not(feature = "http2")))]
-                        let connection = conn::http1::Builder::new()
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service)
-                            .with_upgrades();
-
-                        tokio::pin!(connection);
-
-                        // Serve the connection.
-                        tokio::select! {
-                            result = &mut connection => result.map_err(|e| e.into()),
-                            _ = shutdown_rx.changed() => {
-                                connection.as_mut().graceful_shutdown();
-                                connection.await.map_err(|e| e.into())
-                            }
-                        }
+            // TODO: Attempt to get the task size in the tls example <= 1024.
+            task::spawn(async move {
+                let io = match acceptor.accept(stream).await {
+                    Ok(accepted) => TokioIo::new(accepted),
+                    Err(error) => {
+                        drop(permit);
+                        return Err(error.into());
+                    }
+                };
+
+                // Create a new HTTP/2 connection.
+                #[cfg(feature = "http2")]
+                let connection = conn::http2::Builder::new(TokioExecutor::new())
+                    .timer(TokioTimer::new())
+                    .serve_connection(io, service);
+
+                // Create a new HTTP/1.1 connection.
+                #[cfg(all(feature = "http1", not(feature = "http2")))]
+                let connection = conn::http1::Builder::new()
+                    .timer(TokioTimer::new())
+                    .serve_connection(io, service)
+                    .with_upgrades();
+
+                tokio::pin!(connection);
+
+                // Serve the connection.
+                let result = tokio::select! {
+                    served = &mut connection => served.map_err(|e| e.into()),
+                    _ = shutdown_rx.changed() => {
+                        connection.as_mut().graceful_shutdown();
+                        connection.await.map_err(|e| e.into())
                     }
                 };
 
+                // Explicitly drop the semaphore permit.
                 drop(permit);
+
                 result
-            }
+            })
         });
 
-        for _ in 0..CONNECTION_JOIN_LIMIT {
-            match connections.try_join_next() {
-                Some(result) => joined!(&result),
-                None => break,
+        if semaphore.available_permits() == 0 {
```

Comment:
I really like the idea of requiring that a connection be joined if the server is at capacity.

However, querying the low-level semaphore is probably something we want to avoid.

If we decide to implement this in the future, it's probably best to track this state ourselves.